File: js/semantic.ts
import {
  IStringKeyedObject,
  IViewModel,
  IPseudo,
  IRIVM,
} from "./interfaces.ts";
import { AttrPromise } from "./utils";
import { wrapRustPseudo } from "./pseudos";
import type { PseudoValue } from "./pseudos";
import {
  StaticTile,
  StaticNode,
} from "./static-types";

// Aggregated timing stats for profiling
const timingStats = {
  wasmCalls: 0,
  wasmTotalMs: 0,
  wrapCalls: 0,
  wrapTotalMs: 0,
};

export function resetTimingStats() {
  timingStats.wasmCalls = 0;
  timingStats.wasmTotalMs = 0;
  timingStats.wrapCalls = 0;
  timingStats.wrapTotalMs = 0;
}

export function getTimingStats() {
  return {
    ...timingStats,
    wasmAvgMs: timingStats.wasmCalls > 0 ? timingStats.wasmTotalMs / timingStats.wasmCalls : 0,
    wrapAvgMs: timingStats.wrapCalls > 0 ? timingStats.wrapTotalMs / timingStats.wrapCalls : 0,
  };
}

export function logTimingStats(label: string = '') {
  const stats = getTimingStats();
  console.log(`[timing-stats] ${label} wasm: ${stats.wasmCalls} calls, ${stats.wasmTotalMs.toFixed(1)}ms total (${stats.wasmAvgMs.toFixed(2)}ms avg) | wrap: ${stats.wrapCalls} calls, ${stats.wrapTotalMs.toFixed(1)}ms total`);
}

class SemanticViewModel implements IStringKeyedObject, IViewModel {
  [key: string | symbol]: any;
  _: IViewModel | Promise<IViewModel> | undefined = undefined;
  then: undefined;
  [Symbol.toPrimitive]: undefined;

  get [Symbol.toStringTag]() { return 'SemanticViewModel'; }

  __parentPseudo: PseudoValue<any> | undefined;
  __childValues: Map<string, any>;
  __parentWkri: IRIVM<any> | null;
  __tile: StaticTile | null;
  __node: StaticNode;

  __forJsonCache(): null {
    return null;
  }

  constructor(
    parentWkri: IRIVM<any> | null,
    tile: StaticTile | null,
    node: StaticNode,
  ) {
    this.__childValues = new Map<string, any>();
    this.__parentWkri = parentWkri;
    this.__tile = tile;
    this.__node = node;
    return new Proxy(this, {
      set: (object, key, value) => {
        const k: string = typeof key === 'symbol' ? key.description || '' : key;
        // Allow setting internal properties (starting with __)
        if (k.startsWith("__") || key in object) {
          object[k] = value;
          return true;
        }
        throw Error(`Setting semantic values via proxy (key: ${String(key)}) is not supported`);
      },
      get: (object, key) => {
        const k: string = typeof key === 'symbol' ? key.description || '' : key;

        if (key in object) {
          return object[key];
        } else if (k.startsWith("__") || k in object) {
          return object[k];
        }
        if (k == "length") {
          throw Error("TODO");
        }
        return new AttrPromise((resolve) => {
          const p = object.__get(k);
          p.then(resolve);
        });
      },
    });
  }

  async toString(): Promise<string> {
    const entries = [...this.__childValues.entries()].map(([k, v]) => `${k}: ${v}`);
    return `[[${entries.join(",")}]]`;
  }

  async toObject() {
    if (!this.__parentWkri || !this.__parentWkri.$) {
      return {};
    }

    const wasmWrapper = this.__parentWkri.$.wasmWrapper;
    const model = this.__parentWkri.$.model;

    // Step 1: Check which nodegroups need tiles loaded (sync call)
    const t0 = performance.now();
    const missingNodegroups = wasmWrapper.getMissingNodegroupsForChildren(this.__node.nodeid);
    timingStats.wasmCalls++;
    timingStats.wasmTotalMs += performance.now() - t0;

    // Step 2: Load any missing tiles (async - the only async part)
    if (missingNodegroups.length > 0) {
      const tileLoader = wasmWrapper.getTileLoader();
      if (tileLoader) {
        // Load tiles for each missing nodegroup
        for (const nodegroupId of missingNodegroups) {
          const t1 = performance.now();
          const tiles = await tileLoader(nodegroupId);
          timingStats.wasmTotalMs += performance.now() - t1;
          if (tiles && tiles.length > 0) {
            const t2 = performance.now();
            wasmWrapper.appendTiles(tiles);
            timingStats.wasmCalls++;
            timingStats.wasmTotalMs += performance.now() - t2;
          }
        }
      }
    }

    // Step 3: Get all child values in a single sync batch call (no async boundary crossings)
    const t3 = performance.now();
    const childValuesMap: Map<string, any> = wasmWrapper.getAllSemanticChildValues(
      this.__tile?.tileid || null,
      this.__node.nodeid,
      this.__node.nodegroup_id || null,
    );
    timingStats.wasmCalls++;
    timingStats.wasmTotalMs += performance.now() - t3;

    // Step 4: Wrap all values and build result object
    // Use Promise.all to batch all getValue() calls - reduces promise callback overhead
    const t4 = performance.now();
    const wrappedChildren: Array<[string, any]> = [];
    for (const [alias, rustValue] of childValuesMap.entries()) {
      if (rustValue === null || rustValue === undefined) {
        continue;
      }
      const child = wrapRustPseudo(rustValue, this.__parentWkri, model);
      timingStats.wrapCalls++;
      if (child === null || child === undefined) {
        continue;
      }
      child.parentNode = this.__parentPseudo || null;
      // Cache the wrapped child for future use
      this.__childValues.set(alias, child);
      wrappedChildren.push([alias, child]);
    }

    // Batch all getValue() calls with Promise.all instead of sequential awaits
    const valuePromises = wrappedChildren.map(([alias, child]) =>
      child.getValue().then((value: any) => [alias, value] as [string, any])
    );
    const resolvedValues = await Promise.all(valuePromises);

    const result: Record<string, any> = {};
    for (const [alias, value] of resolvedValues) {
      result[alias] = value;
    }
    timingStats.wrapTotalMs += performance.now() - t4;

    return result;
  }

  async __get(key: string) {
    const childValue = await this.__getChildValue(key);
    if (!childValue) {
      return null;
    }
    return childValue.getValue();
  }

  __has(key: string) {
    const childAliases = this.__parentWkri.$.model.getChildNodeAliases(this.__node.nodeid);
    return childAliases.includes(key);
  }

  async __getChildValue(key: string): Promise<IPseudo> | null | undefined {
    const parent = this.__parentWkri;
    const tile = this.__tile;
    const node = this.__node;
    const wasmWrapper = parent.$.wasmWrapper;
    const childAliases = parent.$.model.getChildNodeAliases(node.nodeid);

    if (!childAliases.includes(key)) {
      throw Error(
        `Semantic node does not have this key: ${key} (${childAliases.join(', ')})`,
      );
    }

    // Check cache first
    if (this.__childValues.has(key)) {
      return this.__childValues.get(key);
    }

    // Try sync path first (avoids async boundary crossing overhead)
    let rustValue: any;
    const t0 = performance.now();
    try {
      // Sync call - returns immediately if tiles loaded, throws if not
      rustValue = wasmWrapper.getSemanticChildValue(
        tile?.tileid || null,
        node.nodeid,
        node.nodegroup_id || null,
        key
      );
      timingStats.wasmCalls++;
      timingStats.wasmTotalMs += performance.now() - t0;
    } catch (e: any) {
      const errorStr = e?.message || String(e);
      if (errorStr.startsWith('TILES_NOT_LOADED:')) {
        // Tiles not loaded - load them and retry
        const nodegroupId = errorStr.split(':')[1].split(' ')[0]; // Extract nodegroup ID
        const tileLoader = wasmWrapper.getTileLoader();
        if (tileLoader) {
          const t1 = performance.now();
          const tiles = await tileLoader(nodegroupId);
          timingStats.wasmTotalMs += performance.now() - t1;
          if (tiles && tiles.length > 0) {
            const t2 = performance.now();
            wasmWrapper.appendTiles(tiles);
            timingStats.wasmCalls++;
            timingStats.wasmTotalMs += performance.now() - t2;
          }
        }
        // Retry sync call after loading tiles
        const t3 = performance.now();
        rustValue = wasmWrapper.getSemanticChildValue(
          tile?.tileid || null,
          node.nodeid,
          node.nodegroup_id || null,
          key
        );
        timingStats.wasmCalls++;
        timingStats.wasmTotalMs += performance.now() - t3;
      } else {
        throw e; // Re-throw non-tile-loading errors
      }
    }

    // Wrap the Rust value - Rust always returns something (possibly empty PseudoList)
    const t4 = performance.now();
    const child = wrapRustPseudo(rustValue, parent, parent.$.model);
    timingStats.wrapCalls++;
    timingStats.wrapTotalMs += performance.now() - t4;

    if (child === null || child === undefined) {
      return child;
    }

    // Set parent node
    child.parentNode = this.__parentPseudo || null;

    // Cache the child
    this.__childValues.set(key, child);

    return child;
  }

  static async __create(
    tile: StaticTile,
    node: StaticNode,
    _value: any,
    parent: IRIVM<any> | null,
  ): Promise<SemanticViewModel> {
    // Note: value parameter is ignored - semantic nodes don't have direct values,
    // their children are loaded lazily via __getChildValue
    return new SemanticViewModel(parent, tile, node);
  }

  async __asTileData(): Promise<[null, any[]]> {
    // Semantic nodes don't have direct tile values - only their children do.
    // Collect relationships from all cached child values.
    const relationships: any[] = [];
    for (const [_, child] of this.__childValues.entries()) {
      if (child && typeof child.getTile === 'function') {
        const [, childRelationships] = await child.getTile();
        relationships.push(...childRelationships);
      }
    }
    return [null, relationships];
  }
}

export { SemanticViewModel };

File: crates/alizarin-python/python/alizarin/semantic.py
"""
SemanticViewModel matching TypeScript semantic.ts

SemanticViewModel represents a semantic (nested) node in the resource tree.
It provides lazy-loading access to child values.
"""

from __future__ import annotations

from typing import (
    Any,
    Awaitable,
    Dict,
    List,
    Optional,
    TYPE_CHECKING,
    Union,
)

if TYPE_CHECKING:
    from .static_types import StaticNode, StaticTile
    from .pseudos import PseudoValue, IPseudo


# =============================================================================
# SemanticViewModel
# =============================================================================

class SemanticViewModel:
    """
    ViewModel for semantic (nested) nodes.

    Matches TypeScript SemanticViewModel class.
    Provides lazy-loading access to child values via attribute access.

    Attributes:
        __parent_wkri: Parent resource instance
        __tile: The tile this semantic node belongs to (may be None for root)
        __node: The StaticNode definition
        __child_values: Cache of loaded child values
        __parent_pseudo: Parent PseudoValue for context
    """

    __slots__ = (
        "__parent_wkri",
        "__tile",
        "__node",
        "__child_values",
        "__parent_pseudo",
        "_",
        "then",
    )

    def __init__(
        self,
        parent_wkri: Optional[Any],
        tile: Optional[StaticTile],
        node: StaticNode,
    ) -> None:
        """
        Initialize SemanticViewModel.

        Args:
            parent_wkri: Parent resource instance (WKRI)
            tile: The tile containing this semantic node
            node: The StaticNode definition
        """
        self.__parent_wkri: Optional[Any] = parent_wkri
        self.__tile: Optional[StaticTile] = tile
        self.__node: StaticNode = node
        self.__child_values: Dict[str, Any] = {}
        self.__parent_pseudo: Optional[PseudoValue[Any]] = None
        self._: Optional[Any] = None  # IViewModel compatibility
        self.then: None = None  # Not thenable

    def __repr__(self) -> str:
        alias = self.__node.alias if self.__node else "?"
        tile_id = self.__tile.tileid if self.__tile else "no-tile"
        return f"<SemanticViewModel {alias} tile={tile_id}>"

    async def __str__(self) -> str:
        """String representation with child values."""
        entries = [f"{k}: {v}" for k, v in self.__child_values.items()]
        return f"[[{', '.join(entries)}]]"

    def __for_json_cache(self) -> None:
        """Semantic nodes don't have cache entries."""
        return None

    def describe_field(self) -> Optional[str]:
        """Get field description from parent pseudo."""
        if self.__parent_pseudo:
            return self.__parent_pseudo.describe_field()
        return None

    def describe_field_group(self) -> Optional[str]:
        """Get field group description from parent pseudo."""
        if self.__parent_pseudo:
            return self.__parent_pseudo.describe_field_group()
        return None

    async def to_object(self) -> Dict[str, Any]:
        """
        Convert semantic node to a dictionary with all child values.

        Matches TypeScript SemanticViewModel.toObject().
        Loads all children and returns them as a dictionary.

        Returns:
            Dictionary mapping alias to value for all children.
        """
        if not self.__parent_wkri or not hasattr(self.__parent_wkri, "_i"):
            return {}

        instance_wrapper = self.__parent_wkri._i
        if not instance_wrapper:
            return {}

        model = instance_wrapper.model
        if not model:
            return {}

        # Get all child node aliases
        child_aliases = model.get_child_node_aliases(self.__node.nodeid)

        result: Dict[str, Any] = {}
        for alias in child_aliases:
            try:
                child_pseudo = await self.__get_child_value(alias)
                if child_pseudo:
                    value = await child_pseudo.get_value()
                    result[alias] = value
            except Exception as e:
                # Log but continue with other children
                print(f"Warning: Failed to load child {alias}: {e}")
                continue

        return result

    async def __get(self, key: str) -> Any:
        """
        Get child value by key.

        Args:
            key: The child node alias

        Returns:
            The child ViewModel value
        """
        child_value = await self.__get_child_value(key)
        if not child_value:
            return None
        return await child_value.get_value()

    def __has(self, key: str) -> bool:
        """
        Check if a child key exists.

        Args:
            key: The child node alias

        Returns:
            True if the key exists as a child
        """
        if not self.__parent_wkri or not hasattr(self.__parent_wkri, "_i"):
            return False

        instance_wrapper = self.__parent_wkri._i
        if not instance_wrapper:
            return False

        model = instance_wrapper.model
        if not model:
            return False

        child_aliases = model.get_child_node_aliases(self.__node.nodeid)
        return key in child_aliases

    async def __get_child_value(self, key: str) -> Optional[IPseudo]:
        """
        Get a child pseudo value by alias.

        Args:
            key: The child node alias

        Returns:
            The child PseudoValue or None
        """
        parent = self.__parent_wkri
        tile = self.__tile
        node = self.__node

        if not parent or not hasattr(parent, "_i"):
            return None

        instance_wrapper = parent._i
        if not instance_wrapper:
            return None

        model = instance_wrapper.model
        if not model:
            return None

        child_aliases = model.get_child_node_aliases(node.nodeid)

        if key not in child_aliases:
            raise ValueError(
                f"Semantic node does not have this key: {key} ({', '.join(child_aliases)})"
            )

        # Check cache first
        if key in self.__child_values:
            return self.__child_values[key]

        # Try to get from instance wrapper
        if hasattr(instance_wrapper, "wasm_wrapper"):
            wasm_wrapper = instance_wrapper.wasm_wrapper
            if wasm_wrapper and hasattr(wasm_wrapper, "get_semantic_child_value"):
                try:
                    rust_value = wasm_wrapper.get_semantic_child_value(
                        tile.tileid if tile else None,
                        node.nodeid,
                        node.nodegroup_id,
                        key,
                    )

                    # Wrap the Rust value
                    from .pseudos import wrap_rust_pseudo
                    child = wrap_rust_pseudo(rust_value, parent, model)

                    if child is not None:
                        # Set parent reference
                        if hasattr(child, "parent_value"):
                            child.parent_value = self.__parent_pseudo

                        # Cache the child
                        self.__child_values[key] = child

                    return child
                except Exception as e:
                    error_str = str(e)
                    if error_str.startswith("TILES_NOT_LOADED:"):
                        # Tiles not loaded - would need to load them
                        # For now, return None
                        return None
                    raise

        # Fall back to retrieve_pseudo if available
        if hasattr(instance_wrapper, "retrieve_pseudo"):
            result = await instance_wrapper.retrieve_pseudo(key)
            if result and len(result) > 0:
                child = result[0]
                if hasattr(child, "parent_value"):
                    child.parent_value = self.__parent_pseudo
                self.__child_values[key] = child
                return child

        return None

    async def __as_tile_data(self) -> tuple[None, List[Any]]:
        """
        Convert to tile data format.

        Semantic nodes don't have direct tile values - only their children do.

        Returns:
            Tuple of (None, list of relationships)
        """
        relationships: List[Any] = []
        for child in self.__child_values.values():
            if child and hasattr(child, "get_tile"):
                _, child_relationships = child.get_tile()
                relationships.extend(child_relationships)
        return (None, relationships)

    async def for_json(self) -> Dict[str, Any]:
        """
        Convert to JSON-serializable format.

        Returns all child values as a dictionary.
        """
        return await self.to_object()

    @staticmethod
    async def _create(
        tile: Optional[StaticTile],
        node: StaticNode,
        _value: Any,
        parent: Optional[Any],
    ) -> SemanticViewModel:
        """
        Factory method to create a SemanticViewModel.

        Args:
            tile: The tile containing this semantic node
            node: The StaticNode definition
            _value: Ignored - semantic nodes don't have direct values
            parent: Parent resource instance

        Returns:
            New SemanticViewModel instance
        """
        # Note: value parameter is ignored - semantic nodes don't have direct values,
        # their children are loaded lazily via __get_child_value
        return SemanticViewModel(parent, tile, node)

    def __getattr__(self, name: str) -> Any:
        """
        Attribute access for child values.

        Allows accessing child nodes via attribute syntax: semantic.child_alias
        """
        # Avoid infinite recursion for private attributes
        if name.startswith("_"):
            raise AttributeError(f"'{type(self).__name__}' object has no attribute '{name}'")

        # Return an awaitable that gets the child value
        import asyncio

        async def get_child() -> Any:
            return await self.__get(name)

        # Return a coroutine that can be awaited
        return get_child()
